gesturecontrol.py

```diff
@@ -7,221 +7,259 @@
-
-			# if the 's' key is selected, we are going to "select" a bounding
-			# box to track
-			if key == ord("s"):
-				# select the bounding box of the object we want to track (make
-				# sure you press ENTER or SPACE after selecting the ROI)
-				initBB = cv2.selectROI("Frame", frame, fromCenter=False,
-					showCrosshair=True)
-
-				# start OpenCV object tracker using the supplied bounding box
-				# coordinates, then start the FPS throughput estimator as well
-				self.tracker.init(frame, initBB)
-				fps = FPS().start()
-
-			# if the `q` key was pressed, break from the loop
-			elif key == ord("q"):
-				break
-
-		# if we are using a webcam, release the pointer
-		vs.stop()
-		self.running = False
-
-		# close all windows
-		cv2.destroyAllWindows()
-
-		if writecsv:
-			self.createCSV(vectorDataSet)
-
-	def playback(self):
-		print("[INFO] Playback random values")
-		self.running = True
-
-		while self.running:
-			time.sleep(0.5)
-			key = cv2.waitKey(1) & 0xFF
-
-			# if the `q` key was pressed, break from the loop
-			if key == ord("q"):
-				# Needs work
-				self.running = False
-				break
-
-			if self.vectorLock.acquire(False):
-				self.vectorX = random.randint(-1,1)
-				self.vectorY = random.randint(-1,1)
-				self.vectorZ = 0.0
-				
-				self.vectorLock.release()
-		cv2.destroyAllWindows()
-
-	# Returns the state of gesture control
-	def checkRunning(self):
-		return self.running
-
-	def createCSV(self, dataSet):
-		with open('GesturePlayBack.csv', 'w', newline ='') as file:
-			writer = csv.writer(file)
-			for set in dataSet:
-				writer.writerow(list(set))
+    def __init__(self, trackerType = "csrt"):
+        self.trackerType = trackerType
+        self.running = False
+        self.vectorLock = threading.Lock()
+
+        # initialize a dictionary that maps strings to their corresponding
+        # OpenCV object tracker implementations
+        OPENCV_OBJECT_TRACKERS = {
+            "csrt": cv2.TrackerCSRT_create,
+            "kcf": cv2.TrackerKCF_create,
+            "boosting": cv2.TrackerBoosting_create,
+            "mil": cv2.TrackerMIL_create,
+            "tld": cv2.TrackerTLD_create,
+            "medianflow": cv2.TrackerMedianFlow_create,
+            "mosse": cv2.TrackerMOSSE_create
+        }
+
+        # grab the appropriate object tracker using our dictionary of
+        # OpenCV object tracker objects
+        self.tracker = OPENCV_OBJECT_TRACKERS[trackerType]()
+        self.vectorX = 0.0
+        self.vectorY = 0.0
+        self.vectorZ = 0.0
+
+    def grabVector(self):
+        vectorX = 0.0
+        vectorY = 0.0
+        vectorZ = 0.0
+
+        # Acquire lock, read the values then release lock
+        if self.vectorLock.acquire():
+            vectorX = self.vectorX
+            vectorY = self.vectorY
+            vectorZ = self.vectorZ
+            self.vectorLock.release()
+
+        return (vectorX, vectorY, vectorZ)
+
+    def updateVectors(self, vectorX, vectorY, vectorZ):
```

Comment:
There are two ways to approach this:
1. Make this function lock-free (i.e only call this function is you can acquire the lock)
2. Make this function blocking (i.e acquiring and release the lock) and return TRUE if the self.vectors are updated. FALSE otherwise

Reply:
I like the second idea better